[HOCS-5145] Adding SQS connectivity with test messages.

sathikmadetech opened the pull request:
Added an SQS Client that uses com.amazonaws.services.sqs.AmazonSQSAsync to send and receive messages to the queue.

SQS Queue configurations are defined in LocalStackConfiguration and SQSConfiguration classes.

Dependent on localstack with SQS 

## Uncommment the below Properties when running locally and comment the other three
# spring.datasource.url=jdbc:sqlserver://localhost:5434;databaseName=LaganECM
# spring.datasource.username=sa
# spring.datasource.password=Sqladmin-1

pass VM options 
-Dspring.profiles.active=development,local

LiamMacP commented on build.gradle:
```diff
@@ -13,14 +16,18 @@ repositories {
+	implementation 'com.github.javafaker:javafaker:1.0.2'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
 	implementation group: 'com.microsoft.sqlserver', name: 'mssql-jdbc', version: '6.1.0.jre8'
 	testImplementation group: 'com.h2database', name: 'h2', version: '1.3.148'
```
You don't have any test packages here - are you adding tests?

sathikmadetech replied:
This library classes are used inside TokenReplacer.java

LiamMacP commented on src/main/java/uk/gov/digital/ho/hocs/client/MessageService.java:
```diff
@@ -0,0 +1,55 @@
+            log.info("Successfully sent {}, random messages.", numMessages);
+        }
+
+        sqsClient.read();
```
Is this useful? You're sending many messages but only triggering a single read?

sathikmadetech replied:
Read was for testing and removed now.

LiamMacP commented on src/main/java/uk/gov/digital/ho/hocs/payload/Replacer.java:
```diff
@@ -0,0 +1,31 @@
+import lombok.Getter;
+
+@Getter
+public class Replacer {
```
Will this be used elsewhere - feels like it should be owned as a static class in TokenReplacer

sathikmadetech replied:
Fixed as advised.

LiamMacP commented on src/main/resources/application.properties:
```diff
@@ -2,6 +2,11 @@ spring.datasource.driverClassName=com.microsoft.sqlserver.jdbc.SQLServerDriver
+# spring.datasource.url=jdbc:sqlserver://localhost:5434;databaseName=LaganECM
+# spring.datasource.username=sa
+# spring.datasource.password=Sqladmin-1
 spring.jpa.show-sql=true
```
This should be off by default and on for locally if you want it

sathikmadetech replied:
Fixed in the yml file configuration

LiamMacP commented on src/main/resources/application.properties:
```diff
@@ -12,4 +17,14 @@ spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.SQLServer2016Dia
+run.config.num-messages=1
+run.config.complaint-type=SOMETHING_ELSE
+
 management.health.probes.enabled=true
```
This would be much more readable as a yaml file.

LiamMacP replied:
This file also needs a new line - there will be an option in your IDE for this by default

sathikmadetech replied:
Fixed

LiamMacP reviewed (changes requested):
A fair number of comments - most need actioning. 

A couple of other points: 

- I'm not seeing any tests in this PR - would be good to get that in sooner rather than later for confidence.
- There is no README or LICENSE - can these be added please

LiamMacP reviewed (commented):
You're going to need to add the logback config too so it goes into Kibana yet.

rjweeks reviewed (changes requested):
Hi @sathikmadetech, when you address Liam's comments you will be adding some tests so make sure the Drone pipeline passes. Thanks.
hocs-cms-data-migrator/19/1/2

sathikmadetech commented on build.gradle:
```diff
@@ -2,8 +2,11 @@ plugins {
 	id 'org.springframework.boot' version '2.6.7'
 	id 'io.spring.dependency-management' version '1.0.11.RELEASE'
 	id 'java'
+	id "io.freefair.lombok" version "6.4.3"
```
Fixed in the latest commit.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+import org.springframework.context.annotation.Profile;
+
+@Configuration
+@Profile({"local"})
```
Fixed in the latest commit.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+@Profile({"local"})
+public class LocalStackConfiguration {
+
+    @Value("${aws.sqs.region}")
```
Fixed in the latest commit.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+    @Primary
+    @Bean
+    public AmazonSQSAsync sqsClient() {
+        String host = String.format("http://%s:4576/", awsHost);
```
Fixed in the latest commit.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+    @Value("${aws.local-host}")
+    private String awsHost;
+
+    private final AWSCredentialsProvider awsCredentialsProvider = new AWSCredentialsProvider() {
```
used new AWSStaticCredentialsProvider(
                        new BasicAWSCredentials(accessKey, secretKey) instead.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+        String host = String.format("http://%s:4576/", awsHost);
+        AwsClientBuilder.EndpointConfiguration endpoint = new AwsClientBuilder.EndpointConfiguration(host, region);
+        return AmazonSQSAsyncClientBuilder.standard()
+                .withClientConfiguration(new ClientConfiguration().withProtocol(Protocol.HTTP))
```
Fixed.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/SQSConfiguration.java:
```diff
@@ -0,0 +1,42 @@
+                               @Value("${aws.sqs.secret-key}") String secretKey,
+                               @Value("${aws.sqs.region}") String region) {
+
+        if (!StringUtils.hasText(accessKey)) {
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/MessageService.java:
```diff
@@ -0,0 +1,55 @@
+            for (int i = 0; i < numMessages; i++) {
+                String fileName = PayloadFile.valueOf(complaintType).getFileName();
+                log.info("Sending {} : {}", complaintType, fileName);
+//                sqsClient.sendMessage("Test Message");
```
Removed.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/MessageService.java:
```diff
@@ -0,0 +1,55 @@
+                String msg= replaceEach(getResourceFileAsString(fileName), replacer.getSearchList(), replacer.getReplaceList());
+                sqsClient.sendMessage(msg);
+            }
+            log.info("Successfully sent {}, {} messages.", numMessages, complaintType);
```
Changed to debug.

sathikmadetech commented on src/main/resources/application.properties:
```diff
@@ -2,6 +2,11 @@ spring.datasource.driverClassName=com.microsoft.sqlserver.jdbc.SQLServerDriver
 spring.datasource.username=${db.username:sa}
 spring.datasource.password=${db.password}
+
+## Uncommment the below Properties when running locally
```
Changed to Local and aws config yml files.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/payload/TokenReplacer.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+
+    public static LocalDate convertToLocalDate(Date date) {
```
Moved to DateHelper class

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/payload/TokenReplacer.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Date;
+import java.util.Locale;
+
+public class TokenReplacer {
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/cms/HocsCmsDataMigrator.java:
```diff
@@ -1,13 +1,29 @@
 
 @SpringBootApplication
-public class HocsCmsDataMigrator {
+@ComponentScan(basePackages = {"com.amazonaws.services.sqs", "uk.gov.digital.ho.hocs"})
```
Fixed.

sathikmadetech commented on build.gradle:
```diff
@@ -2,8 +2,11 @@ plugins {
+apply plugin: 'io.spring.dependency-management'
 group = 'uk.gov.digital.ho.hocs'
 version = '0.0.1-SNAPSHOT'
 sourceCompatibility = '11'
```
Fixed

sathikmadetech commented on build.gradle:
```diff
@@ -13,14 +16,18 @@ repositories {
 
 dependencies {
+	implementation group: 'com.amazonaws', name: 'aws-java-sdk', version: '1.12.137'
+	implementation('org.springframework.boot:spring-boot-starter-validation')
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/SQSConfiguration.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.util.StringUtils;
+
+@Configuration
+@Profile({"sqs"})
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/SQSClient.java:
```diff
@@ -0,0 +1,45 @@
+        log.info("Successfully sent MessageId: {} ,queueURL: {}", sendMessageResult.getMessageId(), queueUrl);
+    }
+    public void read() {
+        ReceiveMessageRequest receive_msg_request = new ReceiveMessageRequest()
```
This was added for Dev/Testing and forgot to remove. Removed the read function.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/SQSClient.java:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    public void sendMessage(String message) {
+        SendMessageRequest send_msg_request = new SendMessageRequest()
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/payload/FileReader.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static String getResourceFileAsString(String fileName) {
+        InputStream is = getResourceFileAsInputStream(fileName);
+        if (is != null) {
```
Fixed

sathikmadetech commented on build.gradle:
```diff
@@ -2,8 +2,11 @@ plugins {
+	id "io.freefair.lombok" version "6.4.3"
 }
-
+apply plugin: 'java'
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/config/AWSCustomConfig.java:
```diff
@@ -0,0 +1,32 @@
+@ConfigurationProperties(prefix = "aws")
+@Getter
+@Setter
+public class AWSCustomConfig {
```
Fixed

sathikmadetech commented on src/main/resources/application.properties:
```diff
@@ -2,6 +2,11 @@ spring.datasource.driverClassName=com.microsoft.sqlserver.jdbc.SQLServerDriver
 spring.datasource.url=jdbc:sqlserver://${db.host:localhost}:${db.port:5434};databaseName=LaganECM
```
For now, yes. We will fix this permanently once we get the data from ATOS

sathikmadetech commented on build.gradle:
```diff
@@ -2,8 +2,11 @@ plugins {
 	id 'org.springframework.boot' version '2.6.7'
 	id 'io.spring.dependency-management' version '1.0.11.RELEASE'
 	id 'java'
+	id "io.freefair.lombok" version "6.4.3"
```
changed to annotationProcessor. Fixed.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/aws/LocalStackConfiguration.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    };
+
+    @Primary
```
Fixed

sathikmadetech commented on build.gradle:
```diff
@@ -13,14 +16,18 @@ repositories {
 	testImplementation group: 'com.h2database', name: 'h2', version: '1.3.148'
 }
 
+
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/SQSClient.java:
```diff
@@ -0,0 +1,45 @@
+    private final AmazonSQSAsync sqs;
+    private final String queueUrl;
+
+    @Autowired
```
Fixed

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/cms/HocsCmsDataMigrator.java:
```diff
@@ -1,13 +1,29 @@
 @SpringBootApplication
-public class HocsCmsDataMigrator {
+@ComponentScan(basePackages = {"com.amazonaws.services.sqs", "uk.gov.digital.ho.hocs"})
+@Slf4j
```
Removed.

sathikmadetech commented on build.gradle:
```diff
@@ -13,14 +16,18 @@ repositories {
+
 tasks.named('test') {
 	useJUnitPlatform()
 }
```
Fixed.

sathikmadetech commented on src/main/java/uk/gov/digital/ho/hocs/client/MessageService.java:
```diff
@@ -0,0 +1,55 @@
+
+    public void startSending() {
+
+        if (!StringUtils.isEmpty(complaintType)) {
```
Fixed.

sathikmadetech commented:
Thanks @LiamMacP for all the comments. I have addressed them as advised. I have removed the unused codes and changed properties and config files to yml files.

LiamMacP commented:
The merge into this branch has caused inconsistencies with main - the GitHub actions are for instance incorrect.
